contracts/src/v0.8/automation/upkeeps/UpkeepBalanceMonitor.sol

```diff
@@ -0,0 +1,232 @@
+  /// @member targetPercentage is the percentage of the upkeep's minBalance to top-up to
+  /// @member maxTopUpAmount is the maximum amount of LINK to top-up an upkeep with
+  struct Config {
+    uint8 maxBatchSize;
+    uint24 minPercentage;
+    uint24 targetPercentage;
+    uint96 maxTopUpAmount;
+  }
+
+  // ================================================================
+  // |                           STORAGE                            |
+  // ================================================================
+
+  LinkTokenInterface private immutable LINK_TOKEN;
+
+  uint256[] private s_watchList;
+  Config private s_config;
+  IAutomationForwarder private s_forwarder;
+
+  // ================================================================
+  // |                         CONSTRUCTOR                          |
+  // ================================================================
+
+  /// @param linkToken the Link token address
+  /// @param config the initial config for the contract
+  constructor(LinkTokenInterface linkToken, Config memory config) ConfirmedOwner(msg.sender) {
+    require(address(linkToken) != address(0));
+    LINK_TOKEN = linkToken;
+    setConfig(config);
+  }
+
+  // ================================================================
+  // |                      CORE FUNCTIONALITY                      |
+  // ================================================================
+
+  /// @notice Gets a list of upkeeps that are underfunded
+  /// @return needsFunding list of underfunded upkeepIDs
+  /// @return topUpAmounts amount to top up each upkeep
+  function getUnderfundedUpkeeps() public view returns (uint256[] memory, uint256[] memory) {
+    uint256 numUpkeeps = s_watchList.length;
+    uint256[] memory needsFunding = new uint256[](numUpkeeps);
+    uint256[] memory topUpAmounts = new uint256[](numUpkeeps);
+    Config memory config = s_config;
+    IAutomationRegistryConsumer registry = getRegistry();
+    uint256 availableFunds = LINK_TOKEN.balanceOf(address(this));
+    uint256 count;
+    uint256 upkeepID;
+    for (uint256 i = 0; i < numUpkeeps; i++) {
+      upkeepID = s_watchList[i];
+      uint96 upkeepBalance = registry.getBalance(upkeepID);
+      uint256 minBalance = uint256(registry.getMinBalance(upkeepID));
+      uint256 topUpThreshold = (minBalance * config.minPercentage) / 100;
+      uint256 topUpAmount = ((minBalance * config.targetPercentage) / 100) - upkeepBalance;
+      if (topUpAmount > config.maxTopUpAmount) {
+        topUpAmount = config.maxTopUpAmount;
+      }
+      if (upkeepBalance <= topUpThreshold && availableFunds >= topUpAmount) {
+        needsFunding[count] = upkeepID;
+        topUpAmounts[count] = topUpAmount;
+        count++;
+        availableFunds -= topUpAmount;
+      }
+      if (count == config.maxBatchSize) {
+        break;
+      }
+    }
+    if (count < numUpkeeps) {
+      assembly {
+        mstore(needsFunding, count)
+        mstore(topUpAmounts, count)
```

Comment:
n00b, what does this mean? we are setting the value of `count` to the address of `topUpAmounts`? isn't `topUpAmounts` of type `uint256[]`?

Reply:
these arrays are created by the size of `config.maxBatchSize`
however, we only have `count` number of underfunded upkeeps
this basically cuts the array size to `count` and saves the gas using inline assembly.